conbench/entities/benchmark_result.py

```diff
@@ -30,43 +31,47 @@
+        postgresql.ARRAY(s.Numeric), default=[]
+    )
 
-    case = relationship("Case", lazy="joined")
+    case: Mapped[Case] = relationship("Case", lazy="joined")
     # optional info at the benchmark level (i.e. information that isn't a tag that should create a separate case, but information that's good to hold around like links to logs)
-    optional_benchmark_info = Nullable(postgresql.JSONB)
+    optional_benchmark_info: Mapped[Optional[dict]] = Nullable(postgresql.JSONB)
     # this info should probably be called something like context-info it's details about the context that are optional | we believe won't impact performance
-    info = relationship("Info", lazy="joined")
-    context = relationship("Context", lazy="joined")
-    run = relationship("Run", lazy="select")
-    unit = Nullable(s.Text)
-    time_unit = Nullable(s.Text)
-    batch_id = Nullable(s.Text)
+    info: Mapped[Info] = relationship("Info", lazy="joined")
+    context: Mapped[Context] = relationship("Context", lazy="joined")
+    run: Mapped[Run] = relationship("Run", lazy="select")
+    unit: Mapped[Optional[str]] = Nullable(s.Text)
+    time_unit: Mapped[Optional[str]] = Nullable(s.Text)
+    batch_id: Mapped[Optional[str]] = Nullable(s.Text)
     # Do not store timezone information in the DB. Instead, follow timezone
     # convention: the application code must make sure that what we store is the
     # user-given time in UTC.
-    timestamp = NotNull(s.DateTime(timezone=False))
-    iterations = Nullable(s.Integer)
-    min = Nullable(s.Numeric, check("min>=0"))
-    max = Nullable(s.Numeric, check("max>=0"))
-    mean = Nullable(s.Numeric, check("mean>=0"))
-    median = Nullable(s.Numeric, check("median>=0"))
-    stdev = Nullable(s.Numeric, check("stdev>=0"))
-    q1 = Nullable(s.Numeric, check("q1>=0"))
-    q3 = Nullable(s.Numeric, check("q3>=0"))
-    iqr = Nullable(s.Numeric, check("iqr>=0"))
-    error = Nullable(postgresql.JSONB)
-    validation = Nullable(postgresql.JSONB)
-    change_annotations = Nullable(postgresql.JSONB)
+    timestamp: Mapped[datetime] = NotNull(s.DateTime(timezone=False))
+    iterations: Mapped[Optional[int]] = Nullable(s.Integer)
+    min: Mapped[Optional[float]] = Nullable(s.Numeric, check("min>=0"))
+    max: Mapped[Optional[float]] = Nullable(s.Numeric, check("max>=0"))
+    mean: Mapped[Optional[float]] = Nullable(s.Numeric, check("mean>=0"))
+    median: Mapped[Optional[float]] = Nullable(s.Numeric, check("median>=0"))
+    stdev: Mapped[Optional[float]] = Nullable(s.Numeric, check("stdev>=0"))
+    q1: Mapped[Optional[float]] = Nullable(s.Numeric, check("q1>=0"))
+    q3: Mapped[Optional[float]] = Nullable(s.Numeric, check("q3>=0"))
+    iqr: Mapped[Optional[float]] = Nullable(s.Numeric, check("iqr>=0"))
+    error: Mapped[Optional[dict]] = Nullable(postgresql.JSONB)
+    validation: Mapped[Optional[dict]] = Nullable(postgresql.JSONB)
+    change_annotations: Mapped[Optional[dict]] = Nullable(postgresql.JSONB)
```

Comment:
:muscle:

Reply:
oh, this reminds me: SQLAlchemy would actually let me do this:
```
change_annotations: Mapped[Optional[dict]] 
```
and it would figure out that it's a nullable JSONB column automatically. But I thought that was a little too "magic". I'd rather be as explicit as we were before.

Reply:
> it would figure out that it's a nullable JSONB column automatically

So SQLAlchemy would do introspection of type annotations at runtime? :)

What's interesting here (for others reading along, I know Austin knows this) is that `postgresql.JSONB` allows for an arbitrary JSON document (with nesting, and all JSON types, including array).

To make this line of code `Mapped[Optional[dict]] = Nullable(postgresql.JSONB)` wholesome and 'correct' there must be schema validation / careful JSON document construction happening that is in alignment with `dict`.

And `dict` is seemingly correct right now based on https://github.com/conbench/conbench/blob/5e626d479ed38ef53e02d102f084df4387f331eb/conbench/entities/benchmark_result.py#L233 and https://github.com/conbench/conbench/blob/5e626d479ed38ef53e02d102f084df4387f331eb/conbench/entities/benchmark_result.py#L587.

In the future I think we want to maybe narrow down the schema (and then also type) a little further, do get a handle on nesting.

Reply:
@austin3dickey looking closer at https://github.com/conbench/conbench/blob/5e626d479ed38ef53e02d102f084df4387f331eb/conbench/entities/benchmark_result.py#L231 I think that we do not store `None`, but an empty dict if user didn't provide anything during BenchmarkResult creation. Can you double-check if this can ever be `None`? Or is this again 'only' for the scenario of DB schema migration?

Reply:
Haha, yes, the main point of my comment was that I chose to opt-out of SQLAlchemy doing introspection of type annotations at runtime. I think that's a little too magic.

You're right that our current JSONB schemas are underdone; they all _should_ represent maps, not arrays. And some of them probably shouldn't allow for nesting.

Reply:
>I think that we do not store None, but an empty dict if user didn't provide anything during BenchmarkResult creation. Can you double-check if this can ever be None? Or is this again 'only' for the scenario of DB schema migration?

Right, this should only be nullable due to DB schema migration. I tried to ensure that all future data is stored as empty dictionaries. But I believe at least the Arrow database has many `None`s in this column still, and it might take a while to update those rows.

Reply:
> You're right that our current JSONB schemas are underdone; they all _should_ represent maps, not arrays. And some of them probably shouldn't allow for nesting.

I'd love to see us lock these down better, both on the db side as well as in the request schemas by which its populated. We do need flexibility in places, but in others we have strong, undocumented opinions about JSON structure which are a nightmare for users trying to adapt results.